src/sentry/static/sentry/app/components/externalIssues/abstractExternalIssueForm.tsx

```diff
@@ -0,0 +1,246 @@
+import React from 'react';
+import {Modal} from 'react-bootstrap';
+import debounce from 'lodash/debounce';
+import * as queryString from 'query-string';
+
+import AsyncComponent from 'app/components/asyncComponent';
+import {tct} from 'app/locale';
+import {IntegrationIssueConfig, IssueConfigField} from 'app/types';
+import FieldFromConfig from 'app/views/settings/components/forms/fieldFromConfig';
+import Form from 'app/views/settings/components/forms/form';
+import {FieldValue} from 'app/views/settings/components/forms/model';
+import {FormField} from 'app/views/settings/projectAlerts/issueEditor/ruleNode';
+
+export type ExternalIssueAction = 'create' | 'link';
+
+type Props = AsyncComponent['props'];
+
+type State = {
+  action: ExternalIssueAction | null;
+  integrationDetails: IntegrationIssueConfig | null;
+  dynamicFieldValues: {[key: string]: FieldValue | null} | null;
+} & AsyncComponent['state'];
+
+const DEBOUNCE_MS = 200;
+
+/**
+ * @abstract
+ */
+export default class AbstractExternalIssueForm<
+  P extends Props = Props,
+  S extends State = State
+> extends AsyncComponent<P, S> {
+  shouldRenderBadRequests = true;
+  getDefaultState(): State {
+    return {
+      ...super.getDefaultState(),
+      action: 'create',
+      // This is derived from integrationDetails when it loads.
+      dynamicFieldValues: null,
+      // This is fetched by AsyncComponent.getEndpoints.
+      integrationDetails: null,
+    };
+  }
+
+  refetchConfig = () => {
+    const {action, dynamicFieldValues} = this.state;
+    const query = {action, ...dynamicFieldValues};
+    const endpoint = this.getEndPointString();
+
+    this.api.request(endpoint, {
+      method: 'GET',
+      query,
+      success: (data, _, jqXHR) => {
+        this.handleRequestSuccess({stateKey: 'integrationDetails', data, jqXHR}, true);
+      },
+      error: error => {
+        this.handleError(error, ['integrationDetails', endpoint, null, null]);
+      },
+    });
+  };
+
+  getConfigName = (): string => {
+    // Explicitly returning a non-interpolated string for clarity.
+    const {action} = this.state;
+    switch (action) {
+      case 'create':
+        return 'createIssueConfig';
+      case 'link':
+        return 'linkIssueConfig';
+      default:
+        throw new Error('illegal action');
+    }
+  };
+
+  getDynamicFields = (
+    integrationDetails?: IntegrationIssueConfig
+  ): {[key: string]: FieldValue | null} => {
+    const config: IssueConfigField[] =
+      (integrationDetails || this.state.integrationDetails || {})[this.getConfigName()] ||
```

Comment:
@mgaeta maybe we can break this up to be more readable.
```suggestion
      const details = integrationDetails || this.state.integrationDetails || {};
      details[this.getConfigName()] ||
```

Reply:
I was avoiding adding adding a variable because I didn't know what to name it besides `integrationDetails` but I figured it out and pushed a commit.